samza-azure/src/main/java/org/apache/samza/system/eventhub/metrics/SamzaHistogram.java

```diff
@@ -1,62 +1,81 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
 
 package org.apache.samza.system.eventhub.metrics;
 
 import com.codahale.metrics.ExponentiallyDecayingReservoir;
 import com.codahale.metrics.Histogram;
 import com.codahale.metrics.Snapshot;
-import org.apache.samza.metrics.Gauge;
-import org.apache.samza.metrics.MetricsRegistry;
-
 import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
 import java.util.function.Function;
 import java.util.stream.Collectors;
+import org.apache.samza.metrics.Gauge;
+import org.apache.samza.metrics.MetricsRegistry;
+import org.apache.samza.system.eventhub.SamzaEventHubClientManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 
 /**
  * Creates a {@link Histogram} metric using {@link ExponentiallyDecayingReservoir}
  * Keeps a {@link Gauge} for each percentile
  */
-public class SamzaHistogram {
+public class SamzaHistogram implements Runnable {
+  private static final Logger LOG = LoggerFactory.getLogger(SamzaEventHubClientManager.class.getName());
   private static final List<Double> DEFAULT_HISTOGRAM_PERCENTILES = Arrays.asList(50D, 99D);
+  private static final int UPDATE_FREQUENCY_MS = 60000; // 1 minute
   private final Histogram histogram;
   private final List<Double> percentiles;
   private final Map<Double, Gauge<Double>> gauges;
+  private final String name;
 
-  public SamzaHistogram(MetricsRegistry registry, String group, String name) {
-    this(registry, group, name, DEFAULT_HISTOGRAM_PERCENTILES);
+  public SamzaHistogram(MetricsRegistry registry, String group, String name, ScheduledExecutorService executorService) {
+    this(registry, group, name, DEFAULT_HISTOGRAM_PERCENTILES, executorService);
   }
 
-  public SamzaHistogram(MetricsRegistry registry, String group, String name, List<Double> percentiles) {
+  public SamzaHistogram(MetricsRegistry registry, String group, String name, List<Double> percentiles,
+      ScheduledExecutorService executorService) {
+    this.name = name;
     this.histogram = new Histogram(new ExponentiallyDecayingReservoir());
     this.percentiles = percentiles;
     this.gauges = this.percentiles.stream()
-            .filter(x -> x > 0 && x <= 100)
-            .collect(
-                    Collectors.toMap(Function.identity(), x -> registry.newGauge(group, name + "_" + String.valueOf(0), 0D)));
+        .filter(x -> x > 0 && x <= 100)
+        .collect(
+            Collectors.toMap(Function.identity(), x -> registry.newGauge(group, name + "_" + String.valueOf(0), 0D)));
+    executorService.schedule(this, UPDATE_FREQUENCY_MS, TimeUnit.MILLISECONDS);
```

Comment:
Why switch to an async update? Is this taking too long in the sync path?

Reply:
Previously the calculation was happening for every update which was pretty expensive, We are now doing it only at regular intervals (1 minute) which is what we need.

Reply:
If getSnapshot() isn't expensive, you can also register a custom Gauge (using MetricsRegistry#newGauge(String, Gauge)) that overrides its getValue to do this calculation when polled. This way the calculation will be on the MetricsReporter thread.

Reply:
I considered this, if you see there can be multiple gauges per Histogram, I thought it is wasteful to do multiple updates for the same Histogram, But i do like that it is simple. Do you know if the visit calls are all in sequence? Let's chat about this.